pkg/core/edge_kb.go

```diff
@@ -0,0 +1,296 @@
+func GetEdgeDetails(kb EdgeKB, source reflect.Type, target reflect.Type) EdgeDetails {
+	return kb[Edge{From: source, To: target}]
+}
+
+// GetEdgesWithSource will return all edges where the source type parameter is the From of the edge
+func GetEdgesWithSource(kb EdgeKB, source reflect.Type) []Edge {
+	result := []Edge{}
+	for edge := range kb {
+		if edge.From == source {
+			result = append(result, edge)
+		}
+	}
+	return result
+}
+
+// GetEdgesWithTarget will return all edges where the target type parameter is the To of the edge
+func GetEdgesWithTarget(kb EdgeKB, target reflect.Type) []Edge {
+	result := []Edge{}
+	for edge := range kb {
+		if edge.To == target {
+			result = append(result, edge)
+		}
+	}
+	return result
+}
+
+// FindPaths takes in a source and destination type and finds all valid paths to get from source to destination.
+//
+// Find paths does a Depth First Search to search through all edges in the knowledge base.
+// The function tracks visited edges to prevent cycles during execution
+// It also checks the ValidDestinations for each edge against the original destination node to ensure that the edge is allowed to be used in the instance of the path generation
+//
+// The method will return all paths found
+func FindPaths(kb EdgeKB, source reflect.Type, dest reflect.Type) [][]Edge {
+	zap.S().Debugf("Finding Paths from %s -> %s", source.String(), dest.String())
+	result := [][]Edge{}
+	visitedEdges := map[reflect.Type]bool{}
+	stack := []Edge{}
+	findPaths(kb, source, dest, stack, visitedEdges, &result)
+	return result
+}
+
+// findPaths performs the recursive calls of the parent FindPath function
+//
+// It works under the assumption that an edge is bidirectional and uses the edges ValidDestinations field to determine when that assumption is incorrect
+func findPaths(kb EdgeKB, source reflect.Type, dest reflect.Type, stack []Edge, visited map[reflect.Type]bool, result *[][]Edge) {
+	visited[source] = true
+	if source == dest {
+		*result = append(*result, stack)
+	} else {
+		for _, e := range GetEdgesWithSource(kb, source) {
+			if e.From == source && !visited[e.To] && isValidForPath(kb, e, dest) {
+				findPaths(kb, e.To, dest, append(stack, e), visited, result)
+			}
+		}
+		for _, e := range GetEdgesWithTarget(kb, source) {
+			if e.To == source && !visited[e.From] && isValidForPath(kb, e, dest) {
+				findPaths(kb, e.From, dest, append(stack, e), visited, result)
+			}
+		}
+	}
+	delete(visited, source)
+}
+
+// isValidForPath determines if an edge is valid for an instance of path generation.
+//
+// The criteria is:
+//   - if there is no expansion function and no ValidDestinations, assume all destinations are valid.
+//   - otherwise check to see if the path generations destination is valid for the edge
+func isValidForPath(kb EdgeKB, edge Edge, dest reflect.Type) bool {
+	edgeDetail := kb[edge]
```

Comment:
is it possible for edgeDetail to be nil here if edge isn't in kb?

Reply:
yeah also the only thing being put into here are valid edges since its used by findPaths so we should be ok